### DISCUSSION TITLE
Using Codespaces with our enterprise license?

### DISCUSSION THREAD

**Discussion by joshuacurtiss** [uillinois-community/uillinois-community.github.io#61] [Q&A] [answered]
Greetings! 

I've been setting up and researching the resources at our disposal with the university-provided GitHub enterprise license, and I'm curious if Codespaces is allowed and if so, how the billing will be handled.

As best I could tell, I was able to create a codespace in our organization and start it up. I only ran it for a few minutes to avoid incurring charges, and wanted to get clarification on how the billing is handled before pursuing plans to use it more.

Thanks for any insight on this.

Josh

**Comment by jtgorman** [comment]
Hi Josh! I think some members of the Github Governance Committee asked Github this a while ago and I don't think we ever got a response. I'm throwing a reminder on my calendar tomorrow to send an email to that group. The biggest risk is probably that it'll just stop working for you if we hit some sort of top-level limit.

So far I don't think there's been a lot of use of codespaces, so not a lot is really known yet.

**Reply by joshuacurtiss** [reply]
Thanks! Seems like Codespaces could be useful for quick developer onboarding in projects that have more complicated setups.  Looking forward to any more insight on costs. Thanks.

**Reply by jtgorman** [reply]
Hi,

Our main go-between the University and Google is out, but I think they're back next week.  I'll let you know what we hear. I'd say in the meantime if it's available experimenting shouldn't hurt. Usually if stuff isn't covered it's not available or it'll hit a limit pretty quick. I'm not aware of anyone using it on campus yet.  A lot of people seem to be using Docker containers for similar stuff. I've been thinking it would be nice for cross-department programming or bringing on students. My group haven't done either of these recently.

**Reply by jtgorman** [reply, marked as answer]
Sorry this took so long but we've gotten some answers from GitHub. Short summary is even though it looks like Codespaces is enabled, it's capped at $0.

Currently we don't get any amount of codespaces service included in the license and budget.  AITS would like to get a better idea of people's possible usage. There is some amount of codespaces given w/ personal accounts that might help with estimating -> https://docs.github.com/en/billing/managing-billing-for-github-codespaces/about-billing-for-github-codespaces.

It gets tricky at the enterprise level though because at the moment there's not really any granularity to codespaces We can set a spend cap for the Enterprise as a whole, but that's it. We can't even disable the service, so that's why it might appear like it's enabled, but you're going to get an error if you try to run anything most likely. We can't set quotas by users or orgs and there's not really mechanisms or manpower within AITS / purchasing for figuring out ad-hoc expenses.

It does seem like it might be possible to have overlapping licenses (people w/ personal or pro using those resources) but it's not really clear to me how to get purchasing to actually pay for those :/.

**Reply by joshuacurtiss** [reply]
No worries, thanks so much for looking into it!  It does let you start up a codespace but I only ran it for a minute for fear of incurring a cost, so possibly it would shut down after a very short period.. 

It's a nicety, not a need, and even a shared remote environment can be set up for "free" by just setting one up on a VM if a department has one. 

And, unless it is a complex server setup or some other reason to not develop locally, we're leaning toward Docker as the solution for consistent developer experiences on local developer workstations.  But it's good to know if this feature is or isn't available if a situation calls for it.

Thank you again so much for your response!

**Reply by jtgorman** [reply]
There are some amount of codespace minutes given to individual accounts for free and it's confusing (at least to me) how github decides to charge what for these services. It's possible that you tapped into your personal minutes. Or GitHub isn't correctly charging for short amount of usage. I suspect though any serious usage for a longer period of time would have just lead to an error.  

Docker's worked pretty well for this type of stuff in the library. I think codespaces would be more attractive for teaching classes, or a lot of onboarding of people who don't know docker, or you don't have a lot of hardware or a way to get hardware quickly.
